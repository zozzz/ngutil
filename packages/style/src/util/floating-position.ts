import {
    AlignHorizontalOpposite,
    Alignment,
    AlignmentInput,
    alignmentNormalize,
    AlignVerticalOpposite
} from "./alignment"
// import type { FloatingPositionOptionsNormalized } from "./position"
import { Dimension, Position, Rect, rectContract, rectExpand, rectMoveOrigin, rectOrigin } from "./rect"
import { Sides, SidesInput, sidesNormalize } from "./sides"

export interface FloatingPositionInput {
    dims: FloatingPositionDims
    options: FloatingPositionOptions
}

export interface FloatingPositionDims {
    content: Dimension
    anchor: Rect
    placement: Rect
}

export interface FloatingPositionOptions {
    content: FloatingPositionContentOptions
    anchor: FloatingPositionAnchorOptions
    placement: FloatingPositionPlacementOptions
    horizontalAlt?: FloatingPositionAltInput
    verticalAlt?: FloatingPositionAltInput
}

export interface FloatingPositionContentOptions {
    link: AlignmentInput
}

export interface FloatingPositionAnchorOptions {
    link: AlignmentInput
    margin?: SidesInput
}

export interface FloatingPositionPlacementOptions {
    padding?: SidesInput
}

export interface FloatingPosition {
    content: {
        rect: Position & Readonly<Dimension>
        link: Alignment
    }
    anchor: {
        readonly rect: Readonly<Rect>
        link: Alignment
    }
    placement: {
        readonly rect: Readonly<Rect>
        readonly padding: Readonly<Sides>
        readonly rectWithPadding: Readonly<Rect>
        area: Readonly<Rect>
    }
    connection: Position
}

interface PlacementArea {
    anchorLink: Alignment
    contentLink: Alignment
    area: FloatingPosition["placement"]["area"]
    connection: Position
}

type FloatingPositionAltConst =
    | "none"
    | "flip"
    | "shift"
    | "flip-shift"
    | "greatest"
    | "greatest-shift"
    | "smallest"
    | "smallest-shift"
export type FloatingPositionAltInput = FloatingPositionAltConst | FloatingPositionAltFunc
export type FloatingPositionAltFunc = (pos: FloatingPosition, axis: FloatingPositionAltAxis) => void
export type FloatingPositionAltNorm = Array<FloatingPositionAltFunc>

export const enum FloatingPositionAltAxis {
    Horizontal = "H",
    Vertical = "V"
}

const ZERO_PADDING = sidesNormalize(0)
const ZERO_CONNECTION = { x: 0, y: 0 }

export function floatingPosition({ dims, options }: FloatingPositionInput): FloatingPosition {
    const padding = options.placement.padding ? sidesNormalize(options.placement.padding) : ZERO_PADDING
    const placement: FloatingPosition["placement"] = {
        rect: dims.placement,
        padding: padding,
        rectWithPadding: rectContract(dims.placement, padding),
        area: { ...dims.placement }
    }

    const anchor: FloatingPosition["anchor"] = {
        rect: options.anchor.margin ? rectExpand(dims.anchor, options.anchor.margin) : dims.anchor,
        link: alignmentNormalize(options.anchor.link)
    }

    const content: FloatingPosition["content"] = {
        rect: { x: 0, y: 0, ...dims.content },
        link: alignmentNormalize(options.content.link)
    }

    const position = { placement, anchor, content, connection: ZERO_CONNECTION }
    const area = placementArea(position, anchor.link, content.link)
    setPlacement(position, area)

    if (options.horizontalAlt) {
        applyAlts(position, FloatingPositionAltAxis.Horizontal, options.horizontalAlt)
    }

    if (options.verticalAlt) {
        applyAlts(position, FloatingPositionAltAxis.Vertical, options.verticalAlt)
    }

    return position
}

function placementArea(pos: FloatingPosition, anchorLink: Alignment, contentLink: Alignment): PlacementArea {
    const connection = rectOrigin(pos.anchor.rect, anchorLink)
    let { x, y } = connection
    const placement = pos.placement.rect
    const placementPad = pos.placement.padding
    let width = 0
    let height = 0

    switch (contentLink.horizontal) {
        case "left":
        case "start":
            width = placement.width - x
            break
        case "center":
            if (anchorLink.horizontal === "center") {
                width = placement.width
                x = placement.x
            } else {
                const leftSize = x - placementPad.left.value
                const rightSize = placement.width - x - placementPad.right.value
                const mw = Math.max(0, Math.min(leftSize, rightSize))
                width = mw * 2
                x -= mw
            }
            break
        case "end":
        case "right":
            width = x - placement.x
            x = placement.x
            break
    }

    switch (contentLink.vertical) {
        case "top":
            height = placement.height - y
            break
        case "middle":
            if (anchorLink.vertical === "middle") {
                height = placement.height
                y = placement.y
            } else {
                const topSize = y - placementPad.top.value
                const bottomSize = placement.height - y - placementPad.bottom.value
                const mh = Math.max(0, Math.min(topSize, bottomSize))
                height = mh * 2
                y -= mh
            }

            break
        case "bottom":
            height = y - placement.y
            y = placement.y
            break
    }

    const constraint = pos.placement.rectWithPadding
    const newX = Math.max(x, constraint.x)
    const newY = Math.max(y, constraint.y)
    const diffX = newX - x
    const diffY = newY - y

    const area = {
        x: newX,
        y: newY,
        width: Math.min(newX + width - diffX, constraint.x + constraint.width) - newX,
        height: Math.min(newY + height - diffY, constraint.y + constraint.height) - newY
    }

    return { area, contentLink, anchorLink, connection }
}

function setPlacement(position: FloatingPosition, { area, anchorLink, contentLink, connection }: PlacementArea): void {
    position.placement.area = area
    position.anchor.link = anchorLink
    position.content.link = contentLink

    const plannedContentRect = rectMoveOrigin(position.content.rect, contentLink, connection)
    position.connection = adjustConnection(area, plannedContentRect, connection)
    position.content.rect = rectMoveOrigin(position.content.rect, contentLink, connection)
}

function adjustConnection(
    placement: FloatingPosition["placement"]["area"],
    content: FloatingPosition["content"]["rect"],
    connection: FloatingPosition["connection"]
) {
    const x = Math.max(placement.x, content.x)
    const y = Math.max(placement.y, content.y)
    const leftDiff = x - content.x
    const topDiff = y - content.y
    const widthDiff = placement.x + placement.width - (x + content.width)
    const heightDiff = placement.y + placement.height - (y + content.height)

    const result = { ...connection }
    connection.x += leftDiff
    connection.y += topDiff

    if (widthDiff < 0) {
        connection.x += widthDiff
    }

    if (heightDiff < 0) {
        connection.y += heightDiff
    }
    return result
}

function applyAlts(pos: FloatingPosition, axis: FloatingPositionAltAxis, alts: FloatingPositionAltInput) {
    const norm = positionAltNormalize(alts)
    for (const alt of norm) {
        alt(pos, axis)
    }
}

export function positionAltNormalize(
    input: FloatingPositionAltInput | FloatingPositionAltNorm
): FloatingPositionAltNorm {
    if (Array.isArray(input)) {
        return input
    } else if (typeof input === "function") {
        return [input]
    } else if (typeof input === "string") {
        return altStringToFns(input)
    } else {
        throw new Error(`Invalid alt position: ${input}`)
    }
}

function altStringToFns(input: FloatingPositionAltConst): FloatingPositionAltNorm {
    switch (input) {
        case "none":
            return [none]
        case "flip":
            return [flip]
        case "shift":
            return [shift]
        case "flip-shift":
            return [flip, shift]
        case "greatest":
            return [greatest]
        case "greatest-shift":
            return [greatest, shift]
        case "smallest":
            return [smallest]
        case "smallest-shift":
            return [smallest, shift]
    }
}

function none(pos: FloatingPosition, axis: FloatingPositionAltAxis): void {}

function flip(pos: FloatingPosition, axis: FloatingPositionAltAxis): void {
    const wh: "width" | "height" = axis === FloatingPositionAltAxis.Horizontal ? "width" : "height"
    if (pos.placement.area[wh] >= pos.content.rect[wh]) {
        return
    }
    return greatest(pos, axis)
}

function shift(pos: FloatingPosition, axis: FloatingPositionAltAxis): void {}

function greatest(pos: FloatingPosition, axis: FloatingPositionAltAxis): void {
    _bySize(pos, axis, (a, b) => a >= b)
}

function smallest(pos: FloatingPosition, axis: FloatingPositionAltAxis): void {
    _bySize(pos, axis, (a, b) => a <= b)
}

function _bySize(pos: FloatingPosition, axis: FloatingPositionAltAxis, cmp: (a: number, b: number) => boolean): void {
    const wh: "width" | "height" = axis === FloatingPositionAltAxis.Horizontal ? "width" : "height"

    const current = pos.placement.area[wh]
    const anchorLink = oppositeLink(axis, pos.anchor.link)
    const contentLink = oppositeLink(axis, pos.content.link)
    const oppositePlacement = placementArea(pos, anchorLink, contentLink)

    if (cmp(current, oppositePlacement.area[wh])) {
        return
    }

    setPlacement(pos, oppositePlacement)
}

function oppositeLink(axis: FloatingPositionAltAxis, link: Alignment): Alignment {
    if (axis === FloatingPositionAltAxis.Horizontal) {
        return { horizontal: AlignHorizontalOpposite[link.horizontal], vertical: link.vertical }
    } else {
        return { horizontal: link.horizontal, vertical: AlignVerticalOpposite[link.vertical] }
    }
}

export function floatingPositionToStyle(pos: Readonly<FloatingPosition>): Partial<CSSStyleDeclaration> {
    const contentRect = pos.content.rect
    const placementRect = pos.placement.rect
    const style: Partial<CSSStyleDeclaration> = {}

    if (pos.content.link.horizontal === "right") {
        style["right"] = `${placementRect.width - (contentRect.x + contentRect.width)}px`
        style["left"] = "auto"
    } else {
        style["left"] = `${contentRect.x}px`
        style["right"] = "auto"
    }

    if (pos.content.link.vertical === "bottom") {
        style["bottom"] = `${placementRect.height - (contentRect.y + contentRect.height)}px`
        style["top"] = "auto"
    } else {
        style["top"] = `${contentRect.y}px`
        style["bottom"] = "auto"
    }

    return style
}
