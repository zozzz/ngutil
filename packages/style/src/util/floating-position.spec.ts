import { floatingPosition, FloatingPositionInput } from "./floating-position"

describe("position-compute", () => {
    it("dialog", () => {
        const input: FloatingPositionInput = {
            dims: {
                anchor: { x: 0, y: 0, width: 500, height: 500 },
                placement: { x: 0, y: 0, width: 500, height: 500 },
                content: { width: 100, height: 200 }
            },
            options: {
                anchor: { link: "center middle" },
                content: { link: "center middle" },
                placement: {}
            }
        }

        const res = floatingPosition(input)
        expect(res).toMatchSnapshot()
    })

    it("flip-right-overflow", () => {
        const input: FloatingPositionInput = {
            dims: {
                anchor: { x: 400, y: 0, width: 100, height: 100 },
                placement: { x: 0, y: 0, width: 500, height: 500 },
                content: { width: 200, height: 100 }
            },
            options: {
                anchor: { link: "right middle" },
                content: { link: "left middle" },
                placement: {},
                horizontalAlt: "flip"
            }
        }

        const res = floatingPosition(input)
        expect(res).toMatchSnapshot()
    })

    it("flip-left-overflow", () => {
        const input: FloatingPositionInput = {
            dims: {
                anchor: { x: 30, y: 30, width: 100, height: 100 },
                placement: { x: 0, y: 0, width: 500, height: 500 },
                content: { width: 100, height: 100 }
            },
            options: {
                anchor: { link: "left top" },
                content: { link: "right top" },
                placement: {},
                horizontalAlt: "flip"
            }
        }

        const res = floatingPosition(input)
        expect(res).toMatchSnapshot()
    })

    it("flip-top-overflow", () => {
        const input: FloatingPositionInput = {
            dims: {
                anchor: { x: 30, y: 30, width: 100, height: 100 },
                placement: { x: 0, y: 0, width: 500, height: 500 },
                content: { width: 100, height: 100 }
            },
            options: {
                anchor: { link: "left top" },
                content: { link: "left bottom" },
                placement: {},
                verticalAlt: "flip"
            }
        }

        const res = floatingPosition(input)
        expect(res).toMatchSnapshot()
    })

    it("flip-bottom-overflow", () => {
        const input: FloatingPositionInput = {
            dims: {
                anchor: { x: 30, y: 450, width: 100, height: 100 },
                placement: { x: 0, y: 0, width: 500, height: 500 },
                content: { width: 100, height: 100 }
            },
            options: {
                anchor: { link: "left bottom" },
                content: { link: "left top" },
                placement: {},
                verticalAlt: "flip"
            }
        }

        const res = floatingPosition(input)
        expect(res).toMatchSnapshot()
    })
})
