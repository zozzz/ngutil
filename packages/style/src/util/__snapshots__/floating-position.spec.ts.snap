// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`position-compute dialog 1`] = `
{
  "anchor": {
    "link": {
      "horizontal": "center",
      "vertical": "middle",
    },
    "rect": {
      "height": 500,
      "width": 500,
      "x": 0,
      "y": 0,
    },
  },
  "connection": {
    "x": 250,
    "y": 250,
  },
  "content": {
    "link": {
      "horizontal": "center",
      "vertical": "middle",
    },
    "rect": {
      "height": 200,
      "width": 100,
      "x": 200,
      "y": 150,
    },
  },
  "placement": {
    "area": {
      "height": 500,
      "width": 500,
      "x": 0,
      "y": 0,
    },
    "padding": {
      "bottom": NumberWithUnit {
        "unit": "px",
        "value": 0,
      },
      "left": NumberWithUnit {
        "unit": "px",
        "value": 0,
      },
      "right": NumberWithUnit {
        "unit": "px",
        "value": 0,
      },
      "top": NumberWithUnit {
        "unit": "px",
        "value": 0,
      },
    },
    "rect": {
      "height": 500,
      "width": 500,
      "x": 0,
      "y": 0,
    },
    "rectWithPadding": {
      "height": 500,
      "width": 500,
      "x": 0,
      "y": 0,
    },
  },
}
`;

exports[`position-compute flip-bottom-overflow 1`] = `
{
  "anchor": {
    "link": {
      "horizontal": "left",
      "vertical": "top",
    },
    "rect": {
      "height": 100,
      "width": 100,
      "x": 30,
      "y": 450,
    },
  },
  "connection": {
    "x": 30,
    "y": 450,
  },
  "content": {
    "link": {
      "horizontal": "left",
      "vertical": "bottom",
    },
    "rect": {
      "height": 100,
      "width": 100,
      "x": 30,
      "y": 350,
    },
  },
  "placement": {
    "area": {
      "height": 450,
      "width": 470,
      "x": 30,
      "y": 0,
    },
    "padding": {
      "bottom": NumberWithUnit {
        "unit": "px",
        "value": 0,
      },
      "left": NumberWithUnit {
        "unit": "px",
        "value": 0,
      },
      "right": NumberWithUnit {
        "unit": "px",
        "value": 0,
      },
      "top": NumberWithUnit {
        "unit": "px",
        "value": 0,
      },
    },
    "rect": {
      "height": 500,
      "width": 500,
      "x": 0,
      "y": 0,
    },
    "rectWithPadding": {
      "height": 500,
      "width": 500,
      "x": 0,
      "y": 0,
    },
  },
}
`;

exports[`position-compute flip-left-overflow 1`] = `
{
  "anchor": {
    "link": {
      "horizontal": "right",
      "vertical": "top",
    },
    "rect": {
      "height": 100,
      "width": 100,
      "x": 30,
      "y": 30,
    },
  },
  "connection": {
    "x": 130,
    "y": 30,
  },
  "content": {
    "link": {
      "horizontal": "left",
      "vertical": "top",
    },
    "rect": {
      "height": 100,
      "width": 100,
      "x": 130,
      "y": 30,
    },
  },
  "placement": {
    "area": {
      "height": 470,
      "width": 370,
      "x": 130,
      "y": 30,
    },
    "padding": {
      "bottom": NumberWithUnit {
        "unit": "px",
        "value": 0,
      },
      "left": NumberWithUnit {
        "unit": "px",
        "value": 0,
      },
      "right": NumberWithUnit {
        "unit": "px",
        "value": 0,
      },
      "top": NumberWithUnit {
        "unit": "px",
        "value": 0,
      },
    },
    "rect": {
      "height": 500,
      "width": 500,
      "x": 0,
      "y": 0,
    },
    "rectWithPadding": {
      "height": 500,
      "width": 500,
      "x": 0,
      "y": 0,
    },
  },
}
`;

exports[`position-compute flip-right-overflow 1`] = `
{
  "anchor": {
    "link": {
      "horizontal": "left",
      "vertical": "middle",
    },
    "rect": {
      "height": 100,
      "width": 100,
      "x": 400,
      "y": 0,
    },
  },
  "connection": {
    "x": 400,
    "y": 50,
  },
  "content": {
    "link": {
      "horizontal": "right",
      "vertical": "middle",
    },
    "rect": {
      "height": 100,
      "width": 200,
      "x": 200,
      "y": 0,
    },
  },
  "placement": {
    "area": {
      "height": 500,
      "width": 400,
      "x": 0,
      "y": 0,
    },
    "padding": {
      "bottom": NumberWithUnit {
        "unit": "px",
        "value": 0,
      },
      "left": NumberWithUnit {
        "unit": "px",
        "value": 0,
      },
      "right": NumberWithUnit {
        "unit": "px",
        "value": 0,
      },
      "top": NumberWithUnit {
        "unit": "px",
        "value": 0,
      },
    },
    "rect": {
      "height": 500,
      "width": 500,
      "x": 0,
      "y": 0,
    },
    "rectWithPadding": {
      "height": 500,
      "width": 500,
      "x": 0,
      "y": 0,
    },
  },
}
`;

exports[`position-compute flip-top-overflow 1`] = `
{
  "anchor": {
    "link": {
      "horizontal": "left",
      "vertical": "bottom",
    },
    "rect": {
      "height": 100,
      "width": 100,
      "x": 30,
      "y": 30,
    },
  },
  "connection": {
    "x": 30,
    "y": 130,
  },
  "content": {
    "link": {
      "horizontal": "left",
      "vertical": "top",
    },
    "rect": {
      "height": 100,
      "width": 100,
      "x": 30,
      "y": 130,
    },
  },
  "placement": {
    "area": {
      "height": 370,
      "width": 470,
      "x": 30,
      "y": 130,
    },
    "padding": {
      "bottom": NumberWithUnit {
        "unit": "px",
        "value": 0,
      },
      "left": NumberWithUnit {
        "unit": "px",
        "value": 0,
      },
      "right": NumberWithUnit {
        "unit": "px",
        "value": 0,
      },
      "top": NumberWithUnit {
        "unit": "px",
        "value": 0,
      },
    },
    "rect": {
      "height": 500,
      "width": 500,
      "x": 0,
      "y": 0,
    },
    "rectWithPadding": {
      "height": 500,
      "width": 500,
      "x": 0,
      "y": 0,
    },
  },
}
`;
